test/cpp/src/702-dram-refresh.cc

```diff
@@ -0,0 +1,101 @@
+    offset += LOG2_BLOCK_SIZE;
+    champsim::address_slice channel_slice{champsim::dynamic_extent{champsim::data::bits{champsim::lg2(DRAM_CHANNELS) + offset}, champsim::data::bits{offset}}, 0};
+    offset += champsim::lg2(DRAM_CHANNELS);
+    champsim::address_slice bank_slice{champsim::dynamic_extent{champsim::data::bits{champsim::lg2(DRAM_BANKS) + offset}, champsim::data::bits{offset}}, packet_num % DRAM_BANKS};
+    offset += champsim::lg2(DRAM_BANKS);
+    champsim::address_slice column_slice{champsim::dynamic_extent{champsim::data::bits{champsim::lg2(DRAM_COLUMNS) + offset}, champsim::data::bits{offset}}, 1};
+    offset += champsim::lg2(DRAM_COLUMNS);
+    champsim::address_slice rank_slice{champsim::dynamic_extent{champsim::data::bits{champsim::lg2(DRAM_RANKS) + offset}, champsim::data::bits{offset}}, packet_num % DRAM_RANKS};
+    offset += champsim::lg2(DRAM_RANKS);
+    champsim::address_slice row_slice{champsim::dynamic_extent{champsim::data::bits{64}, champsim::data::bits{offset}}, packet_num % DRAM_ROWS};
+    r.address = champsim::address{champsim::splice(row_slice, rank_slice, column_slice, bank_slice, channel_slice, block_slice)};
+    r.v_address = champsim::address{};
+    r.instr_id = 0;
+    r.response_requested = false;
+
+    if(r.type == access_type::LOAD)
+        channel->add_rq(r);
+    else
+        channel->add_wq(r);
+}
+
+std::vector<bool> refresh_test(MEMORY_CONTROLLER* uut, champsim::channel* channel_uut, uint64_t refresh_cycles, uint64_t DRAM_CHANNELS, uint64_t DRAM_RANKS, uint64_t DRAM_BANKS, uint64_t DRAM_COLUMNS, uint64_t DRAM_ROWS, champsim::chrono::picoseconds tREF)
+{
+    //a refresh MUST be done. If this is violated, then DRAM spec is violated.
+    
+    //record the refresh status of each bank
+    std::vector<bool> bank_refreshed(DRAM_BANKS,false);
+
+    //advanced current time to first refresh cycle, or test will fail since that is the trigger
+    uut->current_time += tREF;
+    //num of refresh cycles to cover
+    //record whether each refresh cycle was respected
+    std::vector<bool> refresh_done;
+
+    //we will cover the first 40 refreshes
+    uint64_t refresh_cycle = 2;
+    uint64_t cycles = 0;
+    while (uut->current_time < champsim::chrono::clock::time_point{} + (tREF*refresh_cycles))
+    {
+        //generate a random packet
+        generate_packet(channel_uut,(uint64_t)cycles, DRAM_CHANNELS, DRAM_RANKS, DRAM_BANKS, DRAM_COLUMNS, DRAM_ROWS);
+        //operate mem controller
+        uut->_operate();
+        cycles++;
+        //make sure that for every refresh cycle, each bank undergoes refresh at least once
+        std::vector<bool> bank_under_refresh;
+        std::transform(std::begin(uut->channels[0].bank_request),std::end(uut->channels[0].bank_request),std::back_inserter(bank_under_refresh),[](const auto& entry){return(entry.under_refresh);});
```

Comment:
I'll cite [my earlier review](https://github.com/ChampSim/ChampSim/pull/390#discussion_r1759801874) and say that this test should measure externally-visible interfaces. If the refresh behavior is externally visible, you should make a test that causes it to happen. If it is not externally visible, we don't need to be modeling it.

Reply:
It is externally visible and very important for performance modeling, just not in a way I would consider easy to test. We would be talking about hard-coding cycle counts like in 701-dram-scheduler (which I really hate, you basically have to rewrite a significant part of the test if you modify timing values). 
Measuring for increases in packet latencies as the controller operates would be the way to do it.

Reply:
Remember the FR-FCFS scheduler has a significant impact on packet scheduling order when latencies change, ergo everything changes when latencies change.

Reply:
Let me paint it a different way: Say that someone comes along in the distant future and decides that they want to revise the DRAM model and improve it. Say that this person realizes a library exists that models different kinds of DRAM very well, and it's called "spamulator". The intrepid researcher builds a bridge between ChampSim and spamulator and wants to verify that they did it correctly. How should they test that ChampSim sees the refreshes it expects? The spamulator implementation does not include a `channels[0].bank_request[0].under_refresh` member. What is the adventurous poor researcher to do?

Reply:
I think it would be wrong to say ChampSim "sees" any refreshes. Latencies spike, but from outside the memory controller, it cannot determine why. I think this could also apply to packet scheduling, row buffer hits, etc... 
By its construction alone, I don't think you can infer these things easily. Its not like a cache. DRAM has a lot of transient aspects. In ChampSim its doable because we know so much about the internal construction and the DRAM model is primitive. We aren't going into any of these tests blind.

In Ramulator, a verified external simulator, we don't know that information, but also I think its strange to be testing for correctness of an external simulator. If Ramulator says a packet took this long, then it took this long. Thats why Ramulator is being used. If we are going to do that, and act like Ramulator can't be trusted, then why not build a test suite inside ChampSim that fully verifies Ramulator?

I think test efforts for Ramulator should focus on the aspects of the interface between Ramulator and ChampSim we control. Refresh clearly falls outside that, and thus I don't think it can be tested. Packet return seems like a much better way to test the correctness of integration, along with checking internal stats of Ramulator (which I will update the tests to do).

If we are going to measure refresh from the external interface, we can do it with loose constraints but not tight ones.
We can expect that if we burst packets for a full tREF, there will at least 1 packet that have a spike in latency and at least 1 that doesn't. We can expect the difference in latencies to be around (tRP + tRAS) * rows refreshed. That is the best we can do.

For now, I would like to focus on testing refresh for our integrated model. The Ramulator pr can incorporate test modifications for Ramulator.